sympy/core/add.py

```diff
@@ -870,10 +870,51 @@ def _eval_as_leading_term(self, x):
             return expr.as_leading_term(x)
 
         infinite = [t for t in expr.args if t.is_infinite]
+        compute = False
 
-        expr = expr.func(*[t.as_leading_term(x) for t in expr.args]).removeO()
-        if not expr:
-            # simple leading term analysis gave us 0 but we have to send
+        leading_terms = [t.as_leading_term(x) for t in expr.args]
+
+        try:
+            min = Order(leading_terms[0]).getn()
+        except Exception:
+            compute = True
+        else:
+            new_expr = leading_terms[0]
+
+        if compute == False:
+            for term in leading_terms[1:]:
+                try:
+                    order = Order(term).getn()
+                except Exception:
+                    compute = True
+                    break
+                else:
+                    if order < min:
+                        min = order
+                        new_expr = term
+                    elif order == min:
+                        new_expr += term
+
+
+        if compute == False:
+            if not new_expr:
+                compute = True
+            elif new_expr is not S.NaN:
+                if new_expr.is_Add:
+                    final_leading_term = [t for t in new_expr.args]
+                else:
+                    final_leading_term = [new_expr]
+
+                if len(expr.args) != len(final_leading_term):
+                    canceled_term = [t for t in leading_terms if t not in final_leading_term]
```

Comment:
The items of `final_leading_term` may be combinations that do not appear in `leading_terms`. For example, if `leading_terms` is `[x, 2*x]` then `linal_leading_term` is `[3*x]`. Hence cancelled terms cannot be reliably found this way.

Reply:
Yes i totally agree with you, like if `expr=x+2*x*cos(x)`.

Reply:
In the original loop itself i can calculate the sum of all cancelled terms, in my opinion.

Reply:
```
if compute == False:
            for term in leading_terms[1:]:
                try:
                    order = Order(term).getn()
                except Exception:
                    compute = True
                    break
                else:
                    if order < min:
                        min = order
                        new_expr = term
                    elif order == min:
                        new_expr += term

if compute == False:
            for term in leading_terms:
                order = Order(term).getn()
                if order > min:
                    cancelled_terms+=term    
```

The last part which I have added can be easily used to compute the `sum of cancelled leading terms`.

Reply:
I made this change and did run the tests, but no improvement, the same testcases are still failing.

Reply:
Is there some reason why you couldn't return `new_expr` (that is now neither 0 nor nan)?

Reply:
In the previous implementation it wasn't, what you suggest should we return 0 or nan, because before returning that it is trying to compute:` old.compute_leading_term(x)`, if some result is found it might return that.

Reply:
Now `new_expr` is the sum of the leading terms of lowest order. It it is not 0 or nan, then it is the leading term of the input expression. Is there some reason why it could not be returned without further ado? (Without passing to `old.compute_leading_term(x)`.)

Reply:
if `new_expr is not 0 or nan`, then it is not passing to `old.compute_leading_term(x)`, rather it is reaching here in my opinion:

```
elif not expr.is_Add:
            return expr
```
where `expr =new_expr` has already been assigned before.

Reply:
Becasue if `new_expr is not 0 or nan` then` compute=False`, then  `old.compute_leading_term(x) `will never get executed.

Reply:
But what is the purpose of this `if len(expr.args) != len(final_leading_term)` test? Why cannot `new_expr` be returned without it?

Reply:
I have a query is it necessary to compute the sum of cancelled terms?

Reply:
There are two types of cancelled terms, leading terms of lowest order that cancel each other (partially or totally) and leading terms of higher order. `new_expr` is the sum of leading terms of lowest order. There is no need to recompute that sum. It that is not 0 (or nan), then it can be returned and the cancelled leading terms of higher order do not affect it. If that if 0, then the series expansions of all terms have to be computed, whether cancelled or not. So there should be no need to (re)compute the sum of cancelled leading terms.

Reply:
```
def _eval_as_leading_term(self, x):
        from sympy import expand_mul, factor_terms, Order

        old = self

        expr = expand_mul(self)
        if not expr.is_Add:
            return expr.as_leading_term(x)

        infinite = [t for t in expr.args if t.is_infinite]
        compute = False
       
        leading_terms = [t.as_leading_term(x) for t in expr.args]

        try:
            min = Order(leading_terms[0]).getn()
        except Exception:
            compute = True
        else:
            new_expr = leading_terms[0]

        if compute == False:
            for term in leading_terms[1:]:
                try:
                    order = Order(term).getn()
                except Exception:
                    compute = True
                    break
                else:
                    if order < min:
                        min = order
                        new_expr = term
                    elif order == min:
                        new_expr += term
       
            expr = new_expr

        if compute:
            # simple leading term analysis gave us cancelled terms but we have to send
            # back a term, so compute the leading term (via series)
            return old.compute_leading_term(x)
        elif expr is S.NaN:
            return old.func._from_args(infinite)
        elif not expr.is_Add:
            return expr
        else:
            plain = expr.func(*[s for s, _ in expr.extract_leading_order(x)])
            rv = factor_terms(plain, fraction=False)
            rv_simplify = rv.simplify()
            # if it simplifies to an x-free expression, return that;
            # tests don't fail if we don't but it seems nicer to do this
            if x not in rv_simplify.free_symbols:
                if rv_simplify.is_zero and plain.is_zero is not True:
                    return (expr - plain)._eval_as_leading_term(x)
                return rv_simplify
            return rv
```

Will this be a better implementation?

Reply:
The organization of this code looks good but the last part where new_expr is Add seems complicated. Can you give examples where that is needed?

Reply:
Are you talking about this part?
```
else:
            plain = new_expr.func(*[s for s, _ in new_expr.extract_leading_order(x)])
            rv = factor_terms(plain, fraction=False)
            rv_simplify = rv.simplify()
            # if it simplifies to an x-free expression, return that;
            # tests don't fail if we don't but it seems nicer to do this
            if x not in rv_simplify.free_symbols:
                if rv_simplify.is_zero and plain.is_zero is not True:
                    return (new_expr - plain)._eval_as_leading_term(x)
                return rv_simplify
            return rv
```

Reply:
if yes, then I don't have any idea as i have simply copied it from the previous implemenation but since, it states: 
```
 # if it simplifies to an x-free expression, return that;
 # tests don't fail if we don't but it seems nicer to do this
```
In my opinion we can remove this part.

Reply:
```
core\tests\test_expr.py", line 466, in test_leadterm3
    assert (y + z + x).leadterm(x) == (y + z, 0)
    line 3346, in leadterm
    should have been free of %s but got %s""" % (self, x, x, c)))
ValueError:
cannot compute leadterm(x + y + z, x). The coefficient should have
been free of x but got x + y + z
```
```
core\tests\test_exprtools.py:test_gcd_terms _______________
line 216, in test_gcd_terms
    s = (a/(x - alpha)).subs(*rep).series(x, 0, 1)
ValueError:
cannot compute leadterm(-_x**3*alpha3/(-sqrt(5)/2 - 1/2) -
_x**2*alpha2/(-sqrt(5)/2 - 1/2) - _x*alpha1/(-sqrt(5)/2 - 1/2) +
_x/(-sqrt(5)/2 - 1/2), _x). The coefficient should have been free of
_x but got -_x**3*alpha3/(-sqrt(5)/2 - 1/2) - _x**2*alpha2/(-sqrt(5)/2
- 1/2) + _x*(-alpha1/(-sqrt(5)/2 - 1/2) + 1/(-sqrt(5)/2 - 1/2))
```

These are some testcases where it is needed.

Reply:
> ```
> def _eval_as_leading_term(self, x):
>         from sympy import expand_mul, factor_terms, Order
> 
>         old = self
> 
>         expr = expand_mul(self)
>         if not expr.is_Add:
>             return expr.as_leading_term(x)
> 
>         infinite = [t for t in expr.args if t.is_infinite]
>         compute = False
>        
>         leading_terms = [t.as_leading_term(x) for t in expr.args]
> 
>         try:
>             min = Order(leading_terms[0]).getn()
>         except Exception:
>             compute = True
>         else:
>             new_expr = leading_terms[0]
> 
>         if compute == False:
>             for term in leading_terms[1:]:
>                 try:
>                     order = Order(term).getn()
>                 except Exception:
>                     compute = True
>                     break
>                 else:
>                     if order < min:
>                         min = order
>                         new_expr = term
>                     elif order == min:
>                         new_expr += term
>            
> 
>         if compute or not new_expr:
>             # simple leading term analysis gave us cancelled terms but we have to send
>             # back a term, so compute the leading term (via series)
>             return old.compute_leading_term(x)
>         elif new_expr is S.NaN:
>             return old.func._from_args(infinite)
>         elif not new_expr.is_Add:
>             return new_expr
>         else:
>             plain = new_expr.func(*[s for s, _ in new_expr.extract_leading_order(x)])
>             rv = factor_terms(plain, fraction=False)
>             rv_simplify = rv.simplify()
>             # if it simplifies to an x-free expression, return that;
>             # tests don't fail if we don't but it seems nicer to do this
>             if x not in rv_simplify.free_symbols:
>                 if rv_simplify.is_zero and plain.is_zero is not True:
>                     return (new_expr - plain)._eval_as_leading_term(x)
>                 return rv_simplify
>             return rv
> ```
> 
> Will this be a better implementation?

Since there is no addition in the failing test cases, it is safe to assume that this implementation is correct. But what are the required changes to make the failing test cases work? @jksuom

Reply:
> ```
> if compute:
>             # simple leading term analysis gave us cancelled terms but we have to send
>             # back a term, so compute the leading term (via series)
> -->       return old.compute_leading_term(x) 
> ```
> 
> old.compute_leading_term(x) is giving **RecursionError on many testcases**, most of the testcases which are failing are simply due to this error/line.

Please take a note of this @jksuom

Reply:
`test_leadterm3` seems to succeed now.

Reply:
So does `test_gcd_terms`.

Reply:
With this code itself?

Reply:
Shall I update my master to check?

Reply:
Make first the changes to `Order` and then test before updating master.

Reply:
Not much of an improvement, only the mentioned cases are passing, the test which was failing in `bin/test core is still failing`. Moreover `one more test started failing in bin/doctest core`.

Only some more cases passed(not all) in bin/test series.

Reply:
Is this `else` part still there? I was hoping that it would not be needed.

        else:
            plain = new_expr.func(*[s for s, _ in new_expr.extract_leading_order(x)])
            rv = factor_terms(plain, fraction=False)
            rv_simplify = rv.simplify()
            # if it simplifies to an x-free expression, return that;
            # tests don't fail if we don't but it seems nicer to do this
            if x not in rv_simplify.free_symbols:
                if rv_simplify.is_zero and plain.is_zero is not True:
                    return (new_expr - plain)._eval_as_leading_term(x)
                return rv_simplify
            return rv

Reply:
Moreover some test in sympy/limits.py is taking way too much time to compute. Shall I commit the mentioned changes so that it becomes easy for you to review.

Reply:
Not much of an improvement, only the mentioned cases are passing, the test which was failing in `bin/doctest core is still failing`. Moreover `one more test started failing in bin/test core`.

Only some more cases passed(not all) in bin/test series.

Reply:
Even after removing the else part, though it now makes sense to remove it.

Reply:
Can you push now the PR with `Order` fixed and no `else` part?

Reply:
Have committed the new changes